### PR TITLE
Contact push/pull fixes and improvements

### PR DESCRIPTION
**Pull request by mattwire**
- ContactPush: Return an array of contact IDs
- ContactPull: Return an array of AccountContactIDs
- Address - don't overwrite Xero address
-

### CODE REVIEW

File: CRM/Civixero/Contact.php [hunk 1]
```diff
@@ -19,7 +19,7 @@ class CRM_Civixero_Contact extends CRM_Civixero_Base {
    *
    * @throws CRM_Core_Exception
    */
-  public function pull(array $params): void {
+  public function pull(array $params): array {
```
Comment (mattwire) [inline CRM/Civixero/Contact.php:22]: return array of AccountContactIDs

File: CRM/Civixero/Contact.php [hunk 2]
```diff
@@ -151,6 +151,7 @@ public function pull(array $params): void {
       // Since we expect this to wind up in the job log we'll print the errors
       throw new CRM_Core_Exception(E::ts('Not all records were saved') . ': ' . print_r($errors, TRUE), 'incomplete', $errors);
     }
+    return ['AccountContactIDs' => $ids ?? []];
```
Comment (mattwire) [inline CRM/Civixero/Contact.php:154]: return array of AccountContactIDs

File: CRM/Civixero/Contact.php [hunk 3]
```diff
@@ -161,18 +162,23 @@ public function pull(array $params): void {
    * @param array $params
    *  - start_date
    *
-   * @return bool
+   * @return array
    * @throws CRM_Core_Exception
    */
-  public function push(array $params, int $limit = 10): bool {
+  public function push(array $params, int $limit = 10): array {
```
Comment (mattwire) [inline CRM/Civixero/Contact.php:168]: return array of Contact IDs

File: CRM/Civixero/Contact.php [hunk 4]
```diff
@@ -281,7 +293,7 @@ public function push(array $params, int $limit = 10): bool {
       // since we expect this to wind up in the job log we'll print the errors
       throw new CRM_Core_Exception(E::ts('Not all contacts were saved') . print_r($errors, TRUE), 'incomplete', $errors);
     }
-    return TRUE;
+    return $contactIDsPushed ?? [];
```
Comment (mattwire) [inline CRM/Civixero/Contact.php:296]: return array of Contact IDs

File: CRM/Civixero/Contact.php [hunk 5]
```diff
@@ -256,13 +268,13 @@ public function push(array $params, int $limit = 10): bool {
...
-        throw new CRM_Core_Exception('Contact Push aborted due to throttling by Xero' . print_r($errors, TRUE));
-      }
-      catch (CRM_Core_Exception $e) {
+      catch (\Exception $e) {
+        // Note: Using \Exception here as we may get various exception types from the Xero API/SDK
         $errorMessage = E::ts('Failed to push contactID: %1') . $record['contact_id'] . ' (' . $record['accounts_contact_id'] . ' )'
-          . E::ts('Error: ') . $e->getMessage() . print_r($responseErrors ?? [], TRUE)
+          . E::ts('Error: ') . $e->getMessage() . '; '
+          . E::ts('Record: ') . print_r($record,TRUE) . '; '
           . E::ts('Contact Push failed');
```
Comment (mattwire) [inline CRM/Civixero/Contact.php:278]: More detailed debug - really helpful when it goes wrong!

File: CRM/Civixero/Contact.php [hunk 6]
```diff
@@ -256,13 +268,13 @@ public function push(array $params, int $limit = 10): bool {
           ->setValues($record)
           ->addValue('accounts_needs_update', FALSE)
           ->execute();
+        $contactIDsPushed[] = $record['contact_id'];
       }
-      catch (CRM_Civixero_Exception_XeroThrottle $e) {
```
Comment (mattwire) [inline CRM/Civixero/Contact.php:260]: Don't think this is required now  it's moved to API wrapper

File: CRM/Civixero/Contact.php [hunk 7]
```diff
@@ -256,13 +268,13 @@ public function push(array $params, int $limit = 10): bool {
           ->setValues($record)
           ->addValue('accounts_needs_update', FALSE)
           ->execute();
+        $contactIDsPushed[] = $record['contact_id'];
```
Comment (mattwire) [inline CRM/Civixero/Contact.php:271]: return array of Contact IDs

File: CRM/Civixero/Contact.php [hunk 8]
```diff
@@ -190,7 +196,13 @@ public function push(array $params, int $limit = 10): bool {
...
         $contact['email'] = $this->getPreferredEmail($locationTypeToSync, $record['contact_id']);
         $contact['phone'] = $this->getPreferredPhone($locationTypeToSync, $record['contact_id']);
-        $contact += $this->getPreferredAddress($locationTypeToSync, $record['contact_id']);
+        // Address is different to the other location fields because it has multiple fields.
+        // We might return NULL from getPreferredAddress which means "do not sync to Xero".
+        // That way we preserve any partial address that we might have in Xero and it will be synced next time it's pulled to Civi.
+        $contactAddress = $this->getPreferredAddress($locationTypeToSync, $record['contact_id']);
+        if ($contactAddress) {
+          $contact = array_merge($contact, $contactAddress);
+        }
```
Comment (mattwire) [inline CRM/Civixero/Contact.php:205]: Per comment - we had some issues with existing (incomplete) addresses in Xero being overwritten. Passing empty keys overwrites Xero, passing nothing does not overwrite Xero.

File: CRM/Civixero/Contact.php (hunk 6 above)
Comment (eileenmcnaughton) [reply CRM/Civixero/Contact.php:260]: What about when it fails mid loop

File: CRM/Civixero/Contact.php (hunk 2 above)
Comment (eileenmcnaughton) [reply CRM/Civixero/Contact.php:154]: Yeah I'm really not loving this - I've been trying to think of something that would look less odd

File: CRM/Civixero/Contact.php (hunk 8 above)
Comment (eileenmcnaughton) [reply CRM/Civixero/Contact.php:201]: this seems fine

File: CRM/Civixero/Contact.php (hunk 2 above)
Comment (mattwire) [reply CRM/Civixero/Contact.php:154]: I don't really mind what the return looks like, it's just *really* useful to have a list of what the API call touched when troubleshooting.

**Pull request closed without merging**